servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,727 @@
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.servlet.AsyncContext;
+import javax.servlet.DispatcherType;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.SessionTrackingMode;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpUpgradeHandler;
+import javax.servlet.http.Part;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.buffer.Unpooled;
+
+/**
+ * The servlet request.
+ */
+final class DefaultServletHttpRequest implements HttpServletRequest {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultServletHttpRequest.class);
+    private static final Locale[] DEFAULT_LOCALS = { Locale.getDefault() };
+    private static final String RFC1123_DATE = "EEE, dd MMM yyyy HH:mm:ss zzz";
+    private static final SimpleDateFormat[] FORMATS_TEMPLATE = {
+            new SimpleDateFormat(RFC1123_DATE, Locale.ENGLISH),
+            new SimpleDateFormat("EEEEEE, dd-MMM-yy HH:mm:ss zzz", Locale.ENGLISH),
+            new SimpleDateFormat("EEE MMMM d HH:mm:ss yyyy", Locale.ENGLISH)
+    };
+
+    private final ServiceRequestContext serviceRequestContext;
+    private final DefaultServletContext servletContext;
+    private final AggregatedHttpRequest httpRequest;
+    private final DefaultServletInputStream inputStream;
+    private final Map<String, Object> attributeMap = new ConcurrentHashMap<>(16);
+    private final SessionTrackingMode sessionIdSource = SessionTrackingMode.COOKIE;
+
+    private final LinkedMultiValueMap<String, String> parameterMap = new LinkedMultiValueMap<>();
+    private final List<Part> fileUploadList = new ArrayList<>();
+    private final String servletPath;
+    private final String pathInfo;
+    private final String requestURI;
+    private final String characterEncoding;
+    private final Locale[] locales;
+    private final String queryString;
+
+    @Nullable
+    private final Cookie[] cookies;
+
+    //Can't be final because user will decide reader or inputStream is initialize.
+    @Nullable
+    private BufferedReader reader;
+
+    private final Map<String, String[]> unmodifiableParameterMap = new AbstractMap<String, String[]>() {
```

Comment:
please remove this and just return `ImmutableMap.copyOf(parameterMap)` where it used.

Reply:
I think unmodifiableParameterMap not only immutable of parameterMap but also convert parammeter value from String to String[]